agora-test/Spec/Effect/TreasuryWithdrawal.hs

```diff
@@ -10,28 +10,272 @@ module Spec.Effect.TreasuryWithdrawal (tests) where
+genTWECases EffectShouldPass                  = do
+  datum <- genTWEDatum
+  txinfo <- txInfoFromTWEDatum return datum
+  return (datum, txinfo)
+
+classifyTWE :: TWETestInput -> TWETestCases
+classifyTWE ((TreasuryWithdrawalDatum r t), info)
+  | paysToEffect = PaysToEffect
+  | outputsDoNotMatchReceivers = OutputsDoNotMatchReceivers
+  | inputsHaveOtherScriptInput = InputsHaveOtherScriptInput
+  | remaindersDoNotReturnToTreasuries = RemaindersDoNotReturnToTreasuries
+  | otherwise = EffectShouldPass
+  where
+    extractCredVal txout = (addressCredential (txOutAddress txout), txOutValue txout)
+    credentialValuePairs = extractCredVal <$> txInfoOutputs info
+    paysToEffect = elem (ScriptCredential $ validatorHash validator) $ fst . extractCredVal <$> (txInfoOutputs info)
+    outputsDoNotMatchReceivers = not $ and $ fmap (\x -> elem x credentialValuePairs) r
+    inputsHaveOtherScriptInput = or $ (\(fst . extractCredVal . txInInfoResolved -> x) ->
+                                         (not $ elem x t) &&
+                                         x /= (ScriptCredential $ validatorHash validator)
+                                      ) <$> txInfoInputs info
+
+    sumValueOfTreasury x = mconcat $ snd <$> (filter (\(c,_) -> elem c t) $ x)
+    treasuryInputSum = sumValueOfTreasury $ extractCredVal . txInInfoResolved <$> txInfoInputs info
+    treasuryOutputSum = sumValueOfTreasury credentialValuePairs
+    expected = Value.unionWith (-) treasuryInputSum (mconcat $ snd <$> r)
+    remaindersDoNotReturnToTreasuries = treasuryOutputSum /= expected
+    
+shrinkTWE :: TWETestInput -> [TWETestInput]
+shrinkTWE = const []
+
+expectedTWE :: Term s (PBuiltinPair PTreasuryWithdrawalDatum PTxInfo :--> PMaybe PUnit)
+expectedTWE = plam $ \_input -> pcon $ PNothing
+
+opaqueToUnit :: Term s (POpaque :--> PUnit)
+opaqueToUnit = plam $ \_ -> pconstant ()
+
+definitionTWE :: Term s (PBuiltinPair PTreasuryWithdrawalDatum PTxInfo :--> PUnit)
+definitionTWE = plam $ \input -> unTermCont $ do
+  datum <- tclet $ pfstBuiltin # input
+  txinfo <- tclet $ psndBuiltin # input
+
+  let scriptContext =
+        pcon $ PScriptContext $
+          pdcons @"txInfo" # pdata txinfo
+          #$ pdcons @"purpose" # pdata (pconstant $ Spending (TxOutRef "0b2086cbf8b6900f8cb65e012de4516cb66b5cb08a9aaba12a8b88be" 1))
+          # pdnil
+
+  pure $ opaqueToUnit #$ treasuryWithdrawalValidator currSymbol
+    # pforgetData (pdata datum)
+    # pforgetData (pdata (pconstant ()))
+    # scriptContext
+
+propertyTWE :: Property
+propertyTWE = classifiedProperty genTWECases shrinkTWE expectedTWE classifyTWE definitionTWE
+
+{- | Generates "lawful" ScriptContext from given TreasuryWithdrawalDatum.
+Other cases can use this ScriptContext to derive from and develop
+a case specific contexts with generators.
+
+TODO: will this work okay with generators adding and removing
+parts? I don't see particular reason it will not to, but will
+that be a "good" generator?
```

Comment:
Good question. This is the thing that is most confusing about how to go about generating `ScriptContext` in general :sweat_smile:. It may be a good place to start, regardless. Using [plutarch-context-builder](https://github.com/Liqwid-Labs/plutarch-context-builder) on top of that likely will get us to where we want.